### PR TITLE
Exclude throw in TraceVariableSet

### LINKED ISSUE #11530: Set-PSDebug -Trace 2 causes Start-Process -PassThru -Wait to crash upon the process exiting
# Steps to reproduce

```powershell
Set-PSDebug -Trace 2
$proc2 = Start-Process -FilePath sc.exe -PassThru -Wait
```

# Expected behavior

Same behavior as when PSDebug is at level 0; the (exited) process object is stored in `$proc2` and no crash occurs

# Actual behavior

```none
PS C:\Users\jahorto> $proc2 = Start-Process -FilePath sc.exe -PassThru -Wait
DEBUG:    1+  >>>> $proc2 = Start-Process -FilePath sc.exe -PassThru -Wait
DEBUG:     ! CALL function '<ScriptBlock>'
DEBUG:   19+                                         if ( &  >>>> { Set-StrictMode -Version 1; $_.PSMessageDetails } ) {
DEBUG:     ! CALL function '<ScriptBlock>'
DEBUG:   19+                                         if ( & {  >>>> Set-StrictMode -Version 1; $_.PSMessageDetails } ) {
DEBUG:   19+                                         if ( & { Set-StrictMode -Version 1;  >>>> $_.PSMessageDetails } ) {
DEBUG:    1+ &  >>>> { Set-StrictMode -Version 1; $this.Exception.InnerException.PSMessageDetails }
DEBUG:     ! CALL function '<ScriptBlock>'
DEBUG:    1+ & {  >>>> Set-StrictMode -Version 1; $this.Exception.InnerException.PSMessageDetails }
DEBUG:    1+ & { Set-StrictMode -Version 1;  >>>> $this.Exception.InnerException.PSMessageDetails }
DEBUG:    1+ & { Set-StrictMode -Version 1; $this.Exception.InnerException.PSMessageDetails  >>>> }
DEBUG:   19+                                         if ( & { Set-StrictMode -Version 1; $_.PSMessageDetails  >>>> } ) {
DEBUG:   25+                                         $errorCategoryMsg = &  >>>> { Set-StrictMode -Version 1; $_.ErrorCategory_Message }

DEBUG:     ! CALL function '<ScriptBlock>'
DEBUG:   25+                                         $errorCategoryMsg = & {  >>>> Set-StrictMode -Version 1; $_.ErrorCategory_Message }

DEBUG:   25+                                         $errorCategoryMsg = & { Set-StrictMode -Version 1;  >>>> $_.ErrorCategory_Message }

DEBUG:   25+                                         $errorCategoryMsg = & { Set-StrictMode -Version 1; $_.ErrorCategory_Message  >>>> }

DEBUG:   41+                                         $originInfo = &  >>>> { Set-StrictMode -Version 1; $_.OriginInfo } 
DEBUG:     ! CALL function '<ScriptBlock>'
DEBUG:   41+                                         $originInfo = & {  >>>> Set-StrictMode -Version 1; $_.OriginInfo } 
DEBUG:   41+                                         $originInfo = & { Set-StrictMode -Version 1;  >>>> $_.OriginInfo } 
DEBUG:   41+                                         $originInfo = & { Set-StrictMode -Version 1; $_.OriginInfo  >>>> } 
Cannot convert value to type System.String.
At line:1 char:1
+ $proc2 = Start-Process -FilePath sc.exe -PassThru -Wait
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidArgument: (:) [], RuntimeException
+ FullyQualifiedErrorId : InvalidCastFromAnyTypeToString
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.2.3
PSEdition                      Core
GitCommitId                    6.2.3
OS                             Microsoft Windows 10.0.19041
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Fix #11530

TraceVariableSet converts a value to string and if the value.ToString() throw the whole PowerShell expression is unexpectedly terminated. The fix is to catch the exception and return "null" string that is good for trace text output.

## PR Context

In #11530 Process.ToString() throws. Tracking issue for Process.ToString() https://github.com/dotnet/runtime/issues/1520

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by jackhorton** [comment]
I have never made a PR to PowerShell before so perhaps I am missing something, but how is it possible that this can only be tested interactively?

**Comment by vexx32** [comment]
If there is a way to run a test for this (if need be we can run a child pwsh process to verify it doesn't crash when using this functionality) we should test this, absolutely. 👍

**Comment by iSazonov** [comment]
@jackhorton We have an edge case here because ToString methods should never throw but Process.ToString() throws if the process has terminated. Single negative impact is that script can break _only_ if we turn on debug. So it is never break mainstream and we can not fear regression.

**Comment by iSazonov** [comment]
@SteveL-MSFT @daxian-dbw Please look the small PR.

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 1]
```diff
@@ -4081,7 +4081,24 @@ internal void TraceVariableSet(string varName, object value)
...
+                try
+                {
+                    valAsString = PSObject.ToStringParser(_context, value);
+                }
+                catch
+                {
+                    // ToStringParser() can throw if the value is already disposed.
+                    valAsString = "null";
+                }
+            }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/debugger/debugger.cs]: Hmm, I think the exception should be handled within `PSObject.ToStringParser`, where the `ToString()` method is called.

Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs]: In debug/tracer we can mask any exception from ToString() without concern.
In engine code I am not sure that we can blindly do the same. 
Docs say that best practice is "avoid exceptions in ToString() methods". So we should consider this as a bug and we shouldn't mask this.

Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs]: @daxian-dbw Could you please continue?

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 2]
```diff
@@ -4157,7 +4157,24 @@ internal void TraceVariableSet(string varName, object value)
             // because 'ToStringParser' would iterate through the enumerator to get the individual elements, which will
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/debugger/debugger.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.8.95', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.165.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 257
257:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

**Comment by iSazonov** [comment]
@daxian-dbw Could you please continue?

File: src/System.Management.Automation/engine/debugger/debugger.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs:4176]: @daxian-dbw Could you please make a conclusion?

**Comment by daxian-dbw** [comment]
@JamesWTruher Could you please review this PR to see if it's OK to sallow the exception only in `TraceVariableSet` and use string `null` in such case?

File: src/System.Management.Automation/engine/debugger/debugger.cs (hunk 1 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/debugger/debugger.cs:4179]: > Docs say that best practice is "avoid exceptions in ToString() methods". So we should consider this as a bug and we shouldn't mask this.

I agree with this, but I think it means we should fix the exception in `PSObject.ToStringParser()`

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 3]
```diff
@@ -4160,7 +4160,24 @@ internal void TraceVariableSet(string varName, object value)
...
             // make irreversible changes to the enumerator.
             bool isValueAnIEnumerator = PSObject.Base(value) is IEnumerator;
-            string valAsString = isValueAnIEnumerator ? typeof(IEnumerator).Name : PSObject.ToStringParser(_context, value);
+            string valAsString = string.Empty;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/debugger/debugger.cs:4163]: It looks like this is always reassigned below, so the assignment is not needed

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 4]
```diff
@@ -4160,7 +4160,24 @@ internal void TraceVariableSet(string varName, object value)
...
+            if (isValueAnIEnumerator)
+            {
+                valAsString = typeof(IEnumerator).Name;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/debugger/debugger.cs:4166]: Isn't this just `nameof(IEnumerator)`?

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 5]
```diff
@@ -4160,7 +4160,24 @@ internal void TraceVariableSet(string varName, object value)
             // because 'ToStringParser' would iterate through the enumerator to get the individual elements, which will
             // make irreversible changes to the enumerator.
             bool isValueAnIEnumerator = PSObject.Base(value) is IEnumerator;
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/debugger/debugger.cs:4162]: Unless this condition is reused elsewhere, it seems like we could just embed it into the `if (...)` condition directly, rather than assign it to a variable

File: src/System.Management.Automation/engine/debugger/debugger.cs [hunk 6]
```diff
@@ -4160,7 +4160,24 @@ internal void TraceVariableSet(string varName, object value)
...
+            else
+            {
+                try
+                {
+                    valAsString = PSObject.ToStringParser(_context, value);
+                }
+                catch
+                {
+                    // ToStringParser() can throw if the value is already disposed.
+                    valAsString = "null";
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/debugger/debugger.cs:4177]: This isn't really accurate -- the value isn't null, and our assumption is that this code path is reached in the event the object has been disposed (even though we catch any exception here).

Is `"null"` a string with a special meaning for us, or would something like `"<disposed>"` work just as well?

File: src/System.Management.Automation/engine/debugger/debugger.cs (hunk 4 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs:4166]: I think the intention is to get runtime name.

File: src/System.Management.Automation/engine/debugger/debugger.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs:4179]: The PR is only for the trace code. For ToStringParser() we could have an issue with repo steps to discuss. It is not clear how many code paths we need to fix there if need.

File: src/System.Management.Automation/engine/debugger/debugger.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/debugger/debugger.cs:4162]: It would be unrelated change.

**Comment by adityapatwardhan** [comment]
@iSazonov Please resolve merge conflicts
